pkg/upbin/upbin.go

```diff
@@ -0,0 +1,398 @@
+)
+
+// Variable factory for unit testing
+var (
+	notifyFunc           = notify
+	needToUpdateFunc     = needToUpdate
+	wantToUpdateFunc     = wantToUpdate
+	saveLastActivityFunc = saveLastActivity
+	prepareURLFunc       = prepareURL
+	managersPackagesFunc = managersPackages
+	whichFunc            = which
+	downloadFileFunc     = downloadFile
+	replaceFileFunc      = replaceFile
+
+	openConfigFunc = openConfig
+
+	packageManagerExistsFunc  = packageManagerExists
+	installPackageManagerFunc = installPackageManager
+)
+
+func UpdateBin() error {
+	notfy, err := notifyFunc()
+	if err != nil {
+		return err
+	}
+
+	if !notfy {
+		return nil
+	}
+
+	if !needToUpdateFunc() {
+		return nil
+	}
+
+	if !wantToUpdateFunc() {
+		return saveLastActivityFunc()
+	}
+
+	fileURL, err := prepareURLFunc()
+	if err != nil {
+		return err
+	}
+
+	install, err := managersPackagesFunc()
+	if err != nil {
+		return err
+	}
+
+	if install {
+		return nil
+	}
+
+	filePath, _ := whichFunc(azioncli)
+
+	err = downloadFileFunc(filePath, fileURL)
+	if err != nil {
+		return err
+	}
+
+	err = replaceFileFunc(filePath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func GetCurrentVersion() int {
+	n, _ := Format(version.BinVersion)
+	return n
+}
+
+func GetLatestVersion() int {
+	r, _ := git.Clone(memory.NewStorage(), nil, &git.CloneOptions{URL: repoURL})
+
+	tags, _ := r.Tags()
+	t, _ := latestTag(tags)
+
+	n, _ := Format(t)
+	return n
+}
+
+func Format(input string) (int, error) {
+	numberString := ""
+	for _, char := range input {
+		if unicode.IsDigit(char) {
+			numberString += string(char)
+		}
+	}
+
+	number, err := strconv.Atoi(numberString)
+	if err != nil {
+		return 0, err
+	}
+
+	return number, nil
+}
+
+type ReferenceIter interface {
+	ForEach(func(*plumbing.Reference) error) error
+}
+
+// latestTag return value in format refs/tags/v0.10.0
+func latestTag(tags ReferenceIter) (tag string, err error) {
+	var biggerVersionSoFar int = 0
+
+	err = tags.ForEach(func(t *plumbing.Reference) error {
+		tagCurrent := t.Name().String() // return this format "refs/tags/v0.10.0"
+
+		if !strings.Contains(tagCurrent, "dev") {
+			versionParts := strings.Split(tagCurrent, ".")
+
+			major := strings.TrimPrefix(versionParts[0], "refs/tags/v")
+			minor := versionParts[1]
+			patch := versionParts[2]
+
+			current, _ := strconv.Atoi(fmt.Sprintf("%s%s%s", major, minor, patch))
+
+			if current > biggerVersionSoFar {
+				biggerVersionSoFar = current
+				tag = tagCurrent
+			}
+		}
+
+		return err
+	})
+
+	return tag, err
+}
+
+func which(command string) (string, error) {
+	path := os.Getenv("PATH")
+	paths := filepath.SplitList(path)
+
+	for _, dir := range paths {
+		executablePath := filepath.Join(dir, command)
+		_, err := os.Stat(executablePath)
+		if err == nil {
+			return executablePath, nil
+		}
+	}
+
+	return "", fmt.Errorf("command '%s' not found", command)
+}
+
+func GetSystem() system {
+	switch system(runtime.GOOS) {
+	case linux:
```

Comment:
You are checking this here, but you never used the information. 
I think your original idea was to get the OS, so then you can run the correct installation procedure for their OS
e.g.: if you verified they are using mac, you can try to run "brew install aziontech/tap/azioncli", then if it fails, you just replace the binary where it's installed. 
At least, that's what I remember from our previous refinements

Reply:
🐴 🐴 🐴